core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/convertTo.kt

```diff
@@ -17,106 +25,209 @@ import org.jetbrains.kotlinx.dataframe.exceptions.ExcessiveColumnsException
 import org.jetbrains.kotlinx.dataframe.exceptions.TypeConversionException
 import org.jetbrains.kotlinx.dataframe.impl.columns.asAnyFrameColumn
 import org.jetbrains.kotlinx.dataframe.impl.emptyPath
+import org.jetbrains.kotlinx.dataframe.impl.schema.createEmptyColumn
 import org.jetbrains.kotlinx.dataframe.impl.schema.createEmptyDataFrame
 import org.jetbrains.kotlinx.dataframe.impl.schema.extractSchema
 import org.jetbrains.kotlinx.dataframe.impl.schema.render
 import org.jetbrains.kotlinx.dataframe.kind
-import org.jetbrains.kotlinx.dataframe.ncol
 import org.jetbrains.kotlinx.dataframe.schema.ColumnSchema
 import org.jetbrains.kotlinx.dataframe.schema.DataFrameSchema
+import org.jetbrains.kotlinx.dataframe.size
 import kotlin.reflect.KType
 import kotlin.reflect.full.withNullability
 import kotlin.reflect.jvm.jvmErasure
 
-private class Converter(val transform: (Any?) -> Any?, val skipNulls: Boolean)
+private open class Converter(val transform: ConverterScope.(Any?) -> Any?, val skipNulls: Boolean)
 
 private class ConvertSchemaDslImpl<T> : ConvertSchemaDsl<T> {
-    val converters = mutableMapOf<Pair<KType, KType>, Converter>()
+    private val converters: MutableMap<Pair<KType, KType>, Converter> = mutableMapOf()
 
+    private val flexibleConverters: MutableMap<(KType, ColumnSchema) -> Boolean, Converter> = mutableMapOf()
+
+    @Suppress("UNCHECKED_CAST")
     override fun <A, B> convert(from: KType, to: KType, converter: (A) -> B) {
-        converters[from.withNullability(false) to to.withNullability(false)] = Converter(converter as (Any?) -> Any?, !from.isMarkedNullable)
+        converters[from.withNullability(false) to to.withNullability(false)] =
+            Converter({ converter(it as A) }, !from.isMarkedNullable)
     }
 
-    fun getConverter(from: KType, to: KType): Converter? {
-        return converters[from.withNullability(false) to to.withNullability(false)]
+    override fun convertIf(
```

Comment:
Do you have some representative example for this API? Honestly, i can't easily understand recursive list thing :(

Reply:
For `convertIf`? Sure! It's not that difficult. `convert.with` can only support exact type matches, which can be a problem when you cannot define exact types for everything you want to catch. Let's say you want to write a converter that would convert all numbers to strings:
```kotlin
convert<Number>.with { it.toString() }
```
This wouldn't work, since `typeOf<Double>() != typeOf<Number>()`. That's where `convertIf` can come in handy. With this function you can manufacture any type of converter you like and you can use the full capabilities of Kotlin reflect you like:
```kotlin
convertIf({ fromType, toSchema -> 
    fromType.withNullability(false).isSubtypeOf(typeOf<Number>()) &&
        toSchema.type.withNullability(false) == typeOf<String>()
}) {
    it.toString()
}
```

For OpenApi I needed to provide a converter for `DataRow -> Any`, however, these converters only work on top-level types, so if a `DataRow->Any` conversion needs to happen in a `List<List<List<DataFrame<*>>>`, then we need my recursive function which makes `convertIf` accept any of these types and convert it accordingly.